keps/0007-dynamic-audit-control.md

```diff
@@ -0,0 +1,126 @@
+
+* [Dynamic Audit Control](#dynamic-audit-control)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+      * [Dynamic Configuration of Policy](#dynamic-configuration-of-policy)
+      * [Dynamic Configuration of Flags](#dynamic-configuration-of-flags)
+      * [User Stories](#user-stories)
+        * [Story 1](#story-1)
+        * [Story 2](#story-2)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Risks and Mitigations](#risks-and-mitigations)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+We want to allow the advanced auditing features to be dynamically configured. Following in the same vein as [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) we would like to provide a means of configuring the auditing features post cluster provisioning.
+
+## Motivation
+
+The advanced auditing features are a powerfull tool, yet difficult to configure. The configuration requires deep insight into the deployment mechanism of choice and often takes many iterations to configure properly requiring a restart of the apiserver each time. Moreover, the ability to install addon tools that configure and enhance audting is hindered by the overhead in configuration. Such tools frequently run on the cluster requiring future knowledge of how to reach them when the cluster is live. These tools could enhance the security and conformance of the cluster and its applications.
+
+### Goals
+- Provide an api and set of objects to configure the advanced auditing kube-apiserver webhook flags dynamically
+- Allow the current audit policy object to be configured dynamically
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration of Policy
+Dynamic configuration of the existing policy object will be provided through use of a shared informer.
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: Policy
+omitStages:
+  - "RequestReceived"
+rules:
+  - level: <request level>
+    resources:
+    - group: <group>
+      resources: <resources>
+```
+In kube-apiserver today this is only accepted as a file, this KEP would allow it to either be configured as a file on the master node or applied to the `kube-system` namespace as an object and be dynamically configured. This will require the audit API to be registered with the apiserver. If both file and object exist the dynamic object will take precedence.
+
+### Dynamic Configuration of Flags
+There are a number of configuration options that are currently set through runtime flags. This KEP would provide a new API for the dynamic configuration of the webhook parameters.
+```yaml
+apiVersion: audit.k8s.io/v1alpha1
```

Comment:
How does this play with aggregated apiservers? All of those get the same AuditConfig?

Reply:
yeah I believe the idea with aggregated apiservers is that all of their traffic still passes through the main apiserver so it will be audited there.

Reply:
No, it won't. It is passed through, but only proxied. We do not look into the connection. The aggregated apiservers will have to watch these objects and pick the config and rules.

Reply:
I see, is the best path forward to simply document what's needed? It would seem we would also need to enforce that the aggregated apiserver is implementing this properly or it opens a potential security gap. Perhaps a conformance test could be created to provide a guarantee.

Reply:
We will implement this in k8s.io/apiserver. So at least every newer apiserver will pick it up automatically. There might be old binaries around though which miss this support.

Reply:
The aggregated API servers will read the configs, not write or serve them. Similary to admission webhooks today. Every aggregated API server will pick them up.

Reply:
The proposal needs to describe how an aggregated API server knows an `AuditConfiguration` is for it.

Reply:
@caesarxuchao I would assume the aggregates would all adhere to the same dynamic config. If a specific config is wanted for that aggregate it could be configured as a static file on the node. Is there a reason an aggregate would need a separate dynamic object?

Reply:
They could be based on a different k8s.io/apiserver version and hence have slightly different requirements, e.g. using another backend.